src/Features/CSharp/Portable/ConvertForEachToFor/CSharpConvertForEachToForCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,178 @@
+using Microsoft.CodeAnalysis.CSharp.CodeStyle.TypeStyle;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.Formatting;
+using Microsoft.CodeAnalysis.Options;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp.ConvertForEachToFor
+{
+    [ExportCodeRefactoringProvider(LanguageNames.CSharp, Name = nameof(CSharpConvertForEachToForCodeRefactoringProvider)), Shared]
+    internal sealed class CSharpConvertForEachToForCodeRefactoringProvider :
+        AbstractConvertForEachToForCodeRefactoringProvider<ForEachStatementSyntax>
+    {
+        protected override string Title => CSharpFeaturesResources.Convert_foreach_to_for;
+
+        protected override ForEachStatementSyntax GetForEachStatement(TextSpan selection, SyntaxToken token)
+        {
+            var foreachStatement = token.Parent.FirstAncestorOrSelf<ForEachStatementSyntax>();
+            if (foreachStatement == null)
+            {
+                return null;
+            }
+
+            // support refactoring only if caret is in between "foreach" and ")"
+            var scope = TextSpan.FromBounds(foreachStatement.ForEachKeyword.Span.Start, foreachStatement.CloseParenToken.Span.End);
+            if (!scope.IntersectsWith(selection))
+            {
+                return null;
+            }
+
+            // check whether there is any comments between foreach and ) tokens
+            // if they do, we don't support conversion.
+            foreach (var trivia in foreachStatement.DescendantTrivia(n => n == foreachStatement || scope.Contains(n.FullSpan)))
+            {
+                if (trivia.Span.End <= scope.Start ||
+                    scope.End <= trivia.Span.Start)
+                {
+                    continue;
+                }
+
+                if (trivia.Kind() != SyntaxKind.WhitespaceTrivia &&
+                    trivia.Kind() != SyntaxKind.EndOfLineTrivia)
+                {
+                    // we don't know what to do with these comments
+                    return null;
+                }
+            }
+
+            return foreachStatement;
+        }
+
+        protected override bool ValidLocation(ForEachInfo foreachInfo)
+        {
+            if (!foreachInfo.RequireCollectionStatement)
+            {
+                return true;
+            }
+
+            // for now, we don't support converting in embeded statement if 
+            // new local declaration for collection is required.
+            // we can support this by using Introduce local variable service
+            // but the service is not currently written in a way that can be
+            // easily reused here.
+            return foreachInfo.ForEachStatement.Parent.IsKind(SyntaxKind.Block);
+        }
+
+        protected override (SyntaxNode start, SyntaxNode end) GetForEachBody(
+            ForEachStatementSyntax foreachStatement) =>
```

Comment:
nit 1.  this parameter really didn't need to wrap.  it would fit fine on any screen and would look nicer for an expression bodied method.
nit 2.  IDE vastly prefers => (indented) on the next line for multiline expression-bodied members.